### PR TITLE
FateVM version interoperability

### PR DESCRIPTION
**Pull request by cytadela8**
- [x] Fixed vm_version handling on remote call
- [x] Basic test for version interoperability
- [x] Test version handling on call
- [x] Test version handling on return 
- [x] Store handling  - use new store on Iris
- [x] Test store behaviour

### CODE REVIEW

**Comment by gorbak25** [comment]
@cytadela8 Please add a release note in docs/specifying that:
1. Starting from IRIS FATE1 contracts use the new FATE2 garbage collector - contract calls to FATE1 contracts are cheaper(as they would be FATE2 contracts)
2. FATE 2 can call FATE 1, FATE 1 can call FATE 2
3. The behavior of opcodes might depend on the VM version:
- New opcodes can be safely used in FATE2, They aren't available in FATE1
- Some opcodes were bugfixed in FATE2(AENS TTL, return None instead of crashing, String.length, current keyblock hash)

**Comment by gorbak25** [comment]
@cytadela8 Another thing to test for:
- FATE2 remote protected call of a FATE1 contract
- Whether rollback of a failing protected call of FATE1 contracts works

**Comment by cytadela8** [comment]
CI fixed. Documentation added. Commits cleaned.

File: apps/aefate/src/aefa_fate.erl [hunk 1]
```diff
@@ -290,10 +290,11 @@ catch_protected(Err, ES) ->
 %% -----------------------------------------------------------
 
 setup_engine(#{ contract := <<_:256>> = ContractPubkey
-              , code := ByteCode} = Spec, Env) ->
+              , code := ByteCode
+              , vm_version := VMVersion} = Spec, Env) ->
     try aeb_fate_code:deserialize(ByteCode) of
         Code ->
-            Cache = #{ ContractPubkey => Code },
+            Cache = #{ ContractPubkey => {Code,VMVersion} },
```
Comment (gorbak25) [inline apps/aefate/src/aefa_fate.erl:297]:
```suggestion
            Cache = #{ ContractPubkey => {Code, VMVersion} },
```

File: apps/aecontract/test/aecontract_SUITE.erl [hunk 2]
```diff
@@ -6676,6 +6838,267 @@ enter_contract(Contract, S) ->
...
+    ForkHeights   = ?config(fork_heights, Cfg),
+    LimaHeight = maps:get(lima, ForkHeights),
+    IrisHeight = maps:get(iris, ForkHeights),
+    Acc = ?call(new_account, 1000000000000000000000000000000000000000 * aec_test_utils:min_gas_price()),
+    LimaSpec      = #{height => LimaHeight, vm_version => ?VM_FATE_SOPHIA_1, gas => 10000000000},
+    IrisSpec      = #{height => IrisHeight, vm_version => ?VM_FATE_SOPHIA_2, gas => 10000000000},
+    {ok, TesterContract} = compile_local_contract("StorageTester"),
+
+    DoTest = fun() ->
+        Tester = ?call(create_contract_with_code, Acc, TesterContract, {}, LimaSpec),
```
Comment (gorbak25) [inline apps/aecontract/test/aecontract_SUITE.erl:7084]: I don't think this mocks the protocol - essentially both contracts run under the FATE1 GC :shrug:

**Review by gorbak25 — APPROVED** [review]

**Review by radrow — APPROVED** [review]
Looks good to me

**Comment by gorbak25** [comment]
@cytadela8 @radrow Let's wait before merging this - we have some time for IRIS rc-1. The more eyes see this PR the better.

**Comment by cytadela8** [comment]
I fixed a style issue. (missing commit I had locally actually :P)

**Review by gorbak25 — APPROVED** [review]

**Review by velzevur — APPROVED** [review]
Looks good, great work @cytadela8

**Pull request merged**